### PR TITLE
Wz wifip2p socket

### PR DESCRIPTION
**Pull request by wenxi-zeng**
- Replace BroadcastReceiver with Observer pattern to update UI when a service is found. I tried to use OnFragmentInteractionListener, but it brings some code related to StudentCallRoll and makes the MainActivity messy.
- Made connection between GroupOwner and Clients.
- Implement StudentRollCallAction by Extending SocketAction.
- Create an abstract WiFiDirectBroadcastReceiver for handling p2p connection changes.
- Add StudentRollCallService. This service calls SocketAction.execute() to send out message
- Add variables of target ip address and roll call port to SocketAction
- Add createGroup method to addLocalService() in MainActivity. The device which calls addLocalService() is defaulted as group owner.

### CODE REVIEW

**Comment by wenxi-zeng** [comment]
@colegarien @symbyte @Perrofrijole 
You guys can follow the step below to create p2p and socket connection:

p2p connection:

```
- Call MainActivity.connectToPeer() somewhere in your fragment
```

socket connection:

```
- Extend BroadcastReceiver/WiFiDirectBroadcastReceiver class
    1. Override onPeerConnected(). The parameter wifiInfo provides the ip address of target device
    2. Send wifiInfo to your fragment

- Go to your fragment. Register the WiFiDirectBroadcastReceiver you just extended by calling
    1. MainActivity.setupBroadcastReciever( [your BroadcastReceiver] );

- Extend SocketActions/SocketAction

- Create your own Service under Services folder
    1. Call SocketAction.execute somewhere in your service

- Go back to your fragment. 
    1. Start your service somewhere in your fragment 
```

Note: A p2p connection should always be established before making a socket connection.

A confirmation is prompted if it is the first two device are paired. You can also check the p2p connection by going to the system settings -> WiFi -> WiFi Direct.  

The sequence of sending a message through socket is: 
- Connect to peer -> BroadcastReceiver notify connected -> send target id to fragment -> put target id and message to Intent -> start service -> calling SocketAction to execute

**Comment by symbyte** [comment]
I don't think that each of us would need to establish our own p2p connection (which as I understand it acts essentially as a network). We would just need to some means of getting the IP addresses of whatever devices we are trying to contact that are already located on the p2p connection (network) that gets started by roll call, which shouldn't be different for each socket interaction (i.e. sending a question). Doing it the way I propose also maintains the strategy of only interacting with those that have checked in for roll call when sending question, and not forcing students to look for another connection when the teacher wants to send questions.

So in short, we dont need a means of creating our own p2p connections, we just need to be able to access the IP information of those that are already involved in the roll call p2p session.

**Comment by wenxi-zeng** [comment]
Right. In that case, the p2p connection is just optional. The socket connection follows what you proposed. Fragment calls Services, Services calls SocketAction. The reason I add BroadcastReceiver is the ip address comes from it.

**Comment by symbyte** [comment]
Ok, so in this case how do we get the IP addresses that are involved in 
the roll call p2p service so that we can use them in our socket actions?

**Comment by colegarien** [comment]
It may be good for wenxi to be in our discussion later

**Comment by symbyte** [comment]
Yes, I agree. Wenxi will you be available to meet with us on hangouts at 
4:00?

**Comment by wenxi-zeng** [comment]
1. Extend the WiFiDirectBroadcastReceiver class. 
2. Override onPeerConnected(). The parameter "wifiInfo" provides the ip address
3. Send "wifiInfo" to your fragment

In your fragment, you can pass the ip address to your service. The service again pass the ip address to SocketAction. Then you can sent the message out.

FYI: the ip address is given on WiFiDirectBroadcastReceiver once the P2P connection is established

**Comment by wenxi-zeng** [comment]
Ah. Sorry. I have an appointment from 4 to 5. My friend is coming to my place. = =||

**Comment by symbyte** [comment]
Ok, thats no problem.

**Comment by wenxi-zeng** [comment]
My method might not be the best solution, and I haven't fully tested yet. Just let me know if you guys have any other idea. I can change my code accordingly.

**Comment by colegarien** [comment]
I am going to close this PR for now, we can discuss some things after the meeting today.
